javalib/src/main/scala/java/lang/Math.scala

```diff
@@ -152,23 +152,53 @@ object Math {
   @alwaysinline def log1p(a: scala.Double): scala.Double =
     cmath.log1p(a)
 
-  @alwaysinline def max(a: scala.Double, b: scala.Double): scala.Double =
-    if (a.isNaN() || b.isNaN()) Double.NaN else `llvm.maxnum.f64`(a, b)
+  // See Issue #3984 re: simplification via LLVM 'maximum' intrinsic.
+  @inline def max(a: scala.Double, b: scala.Double): scala.Double = {
+    if (a.isNaN() || b.isNaN()) Double.NaN
+    else {
+      val mx = `llvm.maxnum.f64`(a, b)
+      if ((mx != 0.0) || (a != b)) mx
+      else if ((1 / a) == Double.POSITIVE_INFINITY) a // When true, have +0.0
```

Comment:
Instead of having a division, could this use `java.lang.Double.doubleToRawLongBits` like in [`java.lang.Double.compare`](https://github.com/scala-native/scala-native/blob/051fc1bb1b4d0c69f1431faf307616dd40e8355c/javalib/src/main/scala/java/lang/Double.scala#L175-L197), which just calls an intrinsic?

I think then this would simply be
```suggestion
      else if (Double.toRawLongBits(a) == 0) a // When true, have +0.0
```

(same question for the `max(Float, Float)`)

Reply:
JD557  Thank you for the review & suggestion.

I'll let this sit on my backbrain for a day or so.

I believe you are correct in the +0.0 case, both Double & Float, change-for-change.

The tricky bit comes with the -0.0 case.  The division is a well worn and well known (to IEEE 754 victims) test
for -0.0. All my professional life I have been told to avoid division, so it goes against my grain to actually 
use it here.   The division is a machine instruction, albeit a costly one, and probably several instructions
to set it up.  

The current `Double.toRawLongBits` idiom testing for -0.0 is actually correct. I spent several hours
tracing to convince myself of that.  It is not the way I would have written that code.  It uses a '>>' right shift
which carries the sign bit to the right (ending up with a -1) and then does some non-obvious, to me, math
to get the proper compare.

One would have to closely examine the optimized machine instruction sequence and create a nano-benchmark
to tell which sequence executes faster.  At least to me, from the view of verifying correctness in economic time,
 the division approach, once one has a clue about the idiom, takes less developer time.

(editorial note: I believe the "== 0" above would be clearer as "== 0L" but I did not want or have {time, energy}
to go changing code all over the place. This work was two or three  levels deep of bugs from what I was
really trying to accomplish.)

Once the -0.0 case used division, I thought it would be clearer if the +0.0 case was implemented in 
a parallel manner.   I am not adverse to making the change you suggest for the +0.0 cases. It avoids
a division and is probably faster at runtime.  I would probably comment at least one of the -0.0 cases
to explain why it is using the division.  

If I make progress on my current objective, I may go back to see if I can re-work the -0.0 case
to remove the division.  I do not quite trust a "if the bits are not exactly 0L then the value must be -0.0".
The engineer in me would like to actually test the sign bit.   I think by the time that bit of code executes,
-0.0 & 0.0 are the only possibilities, so the "if not 0L else ..." approach would work, even in the -0.0 case,
if I could embrace it.  

This discussion helps and, I think, will lead to better code.

What do you think about the execution time gain versus rapid-readability &  parallelism loss?  
<hr>

As mentioned, I did not trust the existing SN `jl.Double.compare()` code.  Part of that was
the series of events (failing tests I was trying to implement) that brought me to that juncture.
Part of that was looking at the code and not being able to convince myself it was correct
given 5 minutes effort (well, I hate to confess in public, a lot longer and several tries more
than that). 

I think I was also influenced by the Scala.js "code" which, IIRC, used the division, at least for -0.0,
and not the `toRawLongBits()` (probably because, I think, the latter is not available on .js.).
When I am implementing, especially under time pressure, I like to copy from people who 
know more than me.

Reply:
>If I make progress on my current objective, I may go back to see if I can re-work the -0.0 case
to remove the division. I do not quite trust a "if the bits are not exactly 0L then the value must be -0.0".
The engineer in me would like to actually test the sign bit. I think by the time that bit of code executes,
-0.0 & 0.0 are the only possibilities, so the "if not 0L else ..." approach would work, even in the -0.0 case,
if I could embrace it.

Yeah, I agree that testing the sign bit would probably be better.

> Once the -0.0 case used division, I thought it would be clearer if the +0.0 case was implemented in
a parallel manner.

Sorry, I'm not sure what you mean by "parallel" here... I assume you mean the implementation being almost like a dual of each other (and not actually some SIMD related thing).

If so, I think that it's still possible to keep that duality by using `else if (Double.toRawLongBits(a) == 0) a // When true, have +0.0` for the max and `else if (Double.toRawLongBits(a) != 0) a // When true, have -0.0` for the min, as the previous `if ((mx != 0.0) || (a != b)) mx` check should already guarantee that we are only comparing 0 and -0 (I think).

> What do you think about the execution time gain versus rapid-readability & parallelism loss?

To be honest, I don't have a strong opinion, especially considering that this can be fixed in the future by a Clang/LLVM bump. So use the version that you prefer.
As long as the link to the issue is there on the comment, I think it's fine

---

P.S. Actually, as a fun puzzler, I think we might even be able to avoid the last branch!

It does make the code quite cryptic, so I wouldn't do that 😅 , just leaving it here for fun.

```scala
  @inline def max(a: scala.Double, b: scala.Double): scala.Double = {
    if (a.isNaN() || b.isNaN()) Double.NaN
    else {
      val mx = Math.max(a, b)
      if ((mx != 0.0) || (a != b)) mx
      else Double.longBitsToDouble(Double.doubleToRawLongBits(a) & Double.doubleToRawLongBits(b)) // If there's a +0.0, clear all the bits
    }
  }
      
  @inline def min(a: scala.Double, b: scala.Double): scala.Double = {
    if (a.isNaN() || b.isNaN()) Double.NaN
    else {
      val mn = `llvm.minnum.f64`(a, b)
      if ((mn != 0.0) || (a != b)) mn
      else Double.longBitsToDouble(Double.doubleToRawLongBits(a) | Double.doubleToRawLongBits(b)) // If there's a +0.0, fill the bits with the other number
    }
  }
```

Reply:
| Sorry, I'm not sure what you mean by "parallel" here... I assume you mean the implementation being almost like a dual | of each other (and not actually some SIMD related thing).

Yes, "dual". "parallel" is the ghost of my high school English teacher urging raw-thinking & unruly students to use
"parallel paragraphs".

Reply:
JD557,

I incorporated your suggestions and think they do lead to simpler and easier to read-on-the-fly code.
This discussion has helped.

In each method a division and an `Infinity`, each a complexity, have been replaced by a comparison
to (Integer) zero. About as simple as it can get.

I also switched the order of guard tests to be `if ((a != b) || (mn != 0.0)) ` because I think the first
clause will trigger more often than the second. A guess, not a measurement, but I am hoping for
lower amortized cost.

The "== 0L" in the `max` case and "!= 0L" in the `min` case are still discordant for me. I believe
they are correct, so I think I will get over it with time & other responsibilities.  I tried a couple
of alternatives in my scratch work, and none were as good.

To my thinking, code that is going to be replaced (by LLVM intrinsics `maximum` and `minimum`) 
Real Soon Now needs to be correct & presentable but not perfect. I do remain open to improvements.
Bit hacking takes Generations.

Reply:
> It does make the code quite cryptic, so I wouldn't do that 😅 , just leaving it here for fun.

A strange idea of fun ;-)  Bit hacking in Scala! Please keep downwind of my C code, I'd hate
to see what you would do.

In the spirit of the fun:

  The example causes a memory write, which I try to avoid in favor of a memory test & branch, but then again
  I am the one who was willing to use a division and an Infinity.